File: Avail/src/com/avail/server/messages/Message.java
/**
 * Message.java
 * Copyright © 1993-2015, The Avail Foundation, LLC.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of the copyright holder nor the names of the contributors
 *   may be used to endorse or promote products derived from this software
 *   without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package com.avail.server.messages;

import com.avail.server.AvailServer;
import com.avail.server.io.AvailServerChannel;

/**
 * An {@link AvailServer} sends and receives {@code Message}s. A {@code Message}
 * received by the server represents a command from the client, whereas a {@code
 * Message} sent by the server represents a response to a command.
 *
 * @author Todd L Smith &lt;todd@availlang.org&gt;
 */
public class Message
{
	/** The maximum allowed size of a frame. */
	public static final int MAX_SIZE = 1_024_000;

	/** The {@linkplain String content} of the {@linkplain Message message}. */
	private final String content;

	/**
	 * Answer the {@linkplain String content} of the {@linkplain Message
	 * message}.
	 *
	 * @return The message's content.
	 */
	public String content ()
	{
		return content;
	}

	/**
	 * Should the {@linkplain AvailServerChannel channel} be {@linkplain
	 * AvailServerChannel#close() closed} after transmitting this {@linkplain
	 * Message message}?
	 */
	private final boolean closeAfterSending;

	/**
	 * Should the {@linkplain AvailServerChannel channel} be {@linkplain
	 * AvailServerChannel#close() closed} after transmitting this {@linkplain
	 * Message message}?
	 *
	 * @return {@code true} if the channel should be closed after sending,
	 *         {@code false} otherwise.
	 */
	public final boolean closeAfterSending ()
	{
		return closeAfterSending;
	}

	/**
	 * Construct a new {@link Message}.
	 *
	 * @param content
	 *        The {@linkplain String content}.
	 */
	public Message (final String content)
	{
		this.content = content;
		this.closeAfterSending = false;
	}

	/**
	 * Construct a new {@link Message}.
	 *
	 * @param content
	 *        The {@linkplain String content}.
	 * @param closeAfterSending
	 *        {@code true} if the {@linkplain AvailServerChannel channel} should
	 *        be {@linkplain AvailServerChannel#close() closed} after
	 *        transmitting this message.
	 */
	public Message (final String content, final boolean closeAfterSending)
	{
		this.content = content;
		this.closeAfterSending = closeAfterSending;
	}

	@Override
	public String toString ()
	{
		return content;
	}
}

File: avail-server/src/main/kotlin/avail/server/messages/Message.kt
/*
 * Message.kt
 * Copyright © 1993-2021, The Avail Foundation, LLC.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of the copyright holder nor the names of the contributors
 *   may be used to endorse or promote products derived from this software
 *   without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package avail.server.messages

import avail.server.AvailServer
import avail.server.io.AvailServerChannel
import org.availlang.json.JSONWriter
import java.nio.charset.StandardCharsets

/**
 * An [AvailServer] sends and receives `Message`s. A `Message` received by the
 * server represents a command from the client, whereas a `Message` sent by the
 * server represents a response to a command.
 *
 * @author Todd L Smith &lt;todd@availlang.org&gt;
 * @author Richard Arriaga &lt;rich@availlang.org&gt;
 *
 * @property content
 *   The [content][ByteArray] of the [message][Message].
 * @property state
 *   The [AvailServerChannel.ProtocolState] the source/target
 *   [AvailServerChannel] was at when this [Message] was received/created.
 * @property closeAfterSending
 *   Should the [channel][AvailServerChannel] be
 *   [closed][AvailServerChannel.scheduleClose] after transmitting this
 *   [message][Message]?
 *
 * @constructor
 * Construct a new [Message].
 *
 * @param content
 *   The [content][ByteArray].
 * @param state
 *   The [AvailServerChannel.ProtocolState] the source/target
 *   [AvailServerChannel] was at when this `Message` was received/created.
 * @param closeAfterSending
 *   `true` if the [channel][AvailServerChannel] should be
 *   [closed][AvailServerChannel.scheduleClose] after transmitting this message.
 */
class Message constructor(
	val content: ByteArray,
	private val state: AvailServerChannel.ProtocolState,
	val closeAfterSending: Boolean = false)
{
	// TODO Process all message data as bytes and transform it just before
	//  actually using the data when being read.
	val stringContent: String get()
	{
		require(!state.generalBinary)
		{
			"Message.stringContent was called but message was received as " +
			"$state."
		}
		return String(content, StandardCharsets.UTF_8)
	}

	/**
	 * Construct a new [Message].
	 *
	 * @param writer
	 *   The [JSONWriter] that contains the message.
	 * @param state
	 *   The [AvailServerChannel.ProtocolState] the source/target
	 *   [AvailServerChannel] was at when this `Message` was received/created.
	 * @param closeAfterSending
	 *   `true` if the [channel][AvailServerChannel] should be
	 *   [closed][AvailServerChannel.scheduleClose] after transmitting this
	 *   message.
	 */
	constructor(
		writer: JSONWriter,
		state: AvailServerChannel.ProtocolState,
		closeAfterSending: Boolean = false
	): this(writer.toString().toByteArray(), state, closeAfterSending)

	companion object
	{
		/** The maximum allowed size of a frame. */
		const val MAX_SIZE = 1024000
	}
}
